DatadogRUM/Sources/Instrumentation/WatchdogTerminations/WatchdogTerminationReporter.swift

```diff
@@ -10,13 +10,45 @@ import DatadogInternal
 /// Reports Watchdog Termination events to Datadog.
 internal protocol WatchdogTerminationReporting {
     /// Sends the Watchdog Termination event to Datadog.
-    func send()
+    func send(date: Date?, state: WatchdogTerminationAppState, viewEvent: RUMViewEvent)
 }
 
 /// Default implementation of `WatchdogTerminationReporting`.
 internal final class WatchdogTerminationReporter: WatchdogTerminationReporting {
+    /// RUM feature scope.
+    private let featureScope: FeatureScope
+
+    init(featureScope: FeatureScope) {
+        self.featureScope = featureScope
+    }
+
     /// Sends the Watchdog Termination event to Datadog.
-    func send() {
-        DD.logger.error("TODO: WatchdogTerminationReporter.report()")
+    func send(date: Date?, state: WatchdogTerminationAppState, viewEvent: RUMViewEvent) {
+        guard state.trackingConsent == .granted else { // consider the user consent from previous session
+            DD.logger.debug("Skipped sending Watchdog Termination as it was recorded with \(state.trackingConsent) consent")
+            return
+        }
+
+        let errorDate = date ?? Date(timeIntervalSinceReferenceDate: TimeInterval(viewEvent.date))
+
+        DD.logger.debug("Sending Watchdog Termination event")
+        featureScope.eventWriteContext(bypassConsent: true) { context, writer in
+            let builder = FatalErrorBuilder(
+                context: context,
+                error: .watchdogTermination,
+                errorDate: errorDate,
+                errorType: "WatchdogTermination",
+                errorMessage: "The operating system watchdog terminated the application.",
+                errorStack: nil,
+                errorThreads: nil,
+                errorBinaryImages: nil,
+                errorWasTruncated: nil,
+                errorMeta: nil
+            )
+            let error = builder.createRUMError(with: viewEvent)
+            let view = builder.updateRUMViewWithError(viewEvent)
+            writer.write(value: error)
+            writer.write(value: view)
```

Comment:
change-request/ We need to consider event timings - RUM intake doesn't accept events in certain situation. Notably, the [`viewEventAvailabilityThreshold`](https://github.com/DataDog/dd-sdk-ios/blob/380bdf4938519b676d5a1d06bd2ce6928637a0a6/DatadogRUM/Sources/FatalErrorBuilder.swift#L13-L19) defines the last moment when we can issue a view update to previous session.

It should be okay for the product to only support the main flow initially, [like it was done in `FatalAppHangsHandler`](https://github.com/DataDog/dd-sdk-ios/blob/380bdf4938519b676d5a1d06bd2ce6928637a0a6/DatadogRUM/Sources/Instrumentation/AppHangs/FatalAppHangsHandler.swift#L114-L124). We can cover edge-case situations later, ideally by moving this logic to `FatalErrorBuilder` instead of duplicating it between all kind of fatal errors:
- [`RUM.CrashReportReceiver`](https://github.com/DataDog/dd-sdk-ios/blob/develop/DatadogRUM/Sources/Integrations/CrashReportReceiver.swift#L156-L173) - supports all edge cases for sending crashes
- [`FatalAppHangsHandler`](https://github.com/DataDog/dd-sdk-ios/blob/380bdf4938519b676d5a1d06bd2ce6928637a0a6/DatadogRUM/Sources/Instrumentation/AppHangs/FatalAppHangsHandler.swift#L114-L124) - supports only the happy-path for sending fatal hangs
- now `WatchdogTerminationReporter`

Reply:
Can you elaborate how does https://github.com/DataDog/dd-sdk-ios/blob/380bdf4938519b676d5a1d06bd2ce6928637a0a6/DatadogRUM/Sources/Instrumentation/AppHangs/FatalAppHangsHandler.swift#L114-L124 logic helps here? 

What makes it confusing is that `else` part only writes the error but this should be covered with view + error already.

Currently we are sending both view update and error. Which one is taken or rejected, is up to the backend logic.

Reply:
> Which one is taken or rejected, is up to the backend logic.

If we want to challenge existing logic it must be first verified with RUM Ingestion. The decision to send error with no view if 4h lasted since crash was based on inputs from BE team. I found first minimal record of this decision [here (see: PROBLEM 1)](https://docs.google.com/document/d/1u4Yr96NtvESbfltDMvqpCJrzl6Sa5KttmKqLSVtbtqQ/edit?usp=sharing). It is dated 2020, so it might be worth revisiting this topic. If this constraint is no longer there, we can vastly simplify App Hangs and Crash Reporting code.

I consider this a blocker as it introduces a divergence in fatal errors handling (WDT / CR / AH).

Reply:
Next: we should do if-else with 4 hours window (constant) and unblock this PR and continue the conversation with RUM BE if we can simplify this.